File: interp.js
"use strict";


const Ops = {
  LEFT: 0,
  RIGHT: 1,
  ADD: 2,
  SUB: 3,
  LOOP_START: 4,
  LOOP_END: 5,
  OUTPUT: 6,
  INPUT: 7,
  ZERO: 8,
  MULTIPLY: 9,
};

class Op {
  constructor(type, arg = 0, offset = 0) {
    this.type = type;
    this.arg = arg;
    this.offset = offset;
  }
}

/* Contraction helper for create_program */
const consec_op = (bytes, i, type) => {
  const start = i;
  while (i < bytes.length && bytes[i] === type) i++;
  const count = i - start;
  return [i - 1, count];
};

const create_program = (bytes) => {
  const prog = [];
  let i = 0;
  while (i < bytes.length) {
    let count = 0;
    switch (bytes[i]) {
      case 60: [i, count] = consec_op(bytes, i, 60); prog.push(new Op(Ops.LEFT, count)); break;
      case 62: [i, count] = consec_op(bytes, i, 62); prog.push(new Op(Ops.RIGHT, count)); break;
      case 43: [i, count] = consec_op(bytes, i, 43); prog.push(new Op(Ops.ADD, count)); break;
      case 45: [i, count] = consec_op(bytes, i, 45); prog.push(new Op(Ops.SUB, count)); break;
      case 91: prog.push(new Op(Ops.LOOP_START)); break;
      case 93: prog.push(new Op(Ops.LOOP_END)); break;
      case 46: prog.push(new Op(Ops.OUTPUT)); break;
      case 44: prog.push(new Op(Ops.INPUT)); break;
      default: break;
    }
    i++;
  }
  return prog;
};

/* Net opposing op counts */
const optimize_contraction = (prog) => {
  const out = [];
  for (const op of prog) {
    // handle data ops (+/-)
    if (op.type === Ops.ADD || op.type === Ops.SUB) {
      // signed count: + for ADD, – for SUB
      let net = (op.type === Ops.ADD ? +1 : -1) * op.arg;
      // if previous was also ADD/SUB, absorb it
      const last = out[out.length - 1];
      if (last && (last.type === Ops.ADD || last.type === Ops.SUB)) {
        out.pop();
        net += (last.type === Ops.ADD ? +1 : -1) * last.arg;
      }
      // only emit if nonzero
      if (net > 0)    out.push(new Op(Ops.ADD, net));
      else if (net < 0) out.push(new Op(Ops.SUB, -net));
      // if net === 0, they cancel entirely
    }
    // handle pointer ops (<, >)
    else if (op.type === Ops.LEFT || op.type === Ops.RIGHT) {
      let net = (op.type === Ops.RIGHT ? +1 : -1) * op.arg;
      const last = out[out.length - 1];
      if (last && (last.type === Ops.LEFT || last.type === Ops.RIGHT)) {
        out.pop();
        net += (last.type === Ops.RIGHT ? +1 : -1) * last.arg;
      }
      if (net > 0)    out.push(new Op(Ops.RIGHT, net));
      else if (net < 0) out.push(new Op(Ops.LEFT, -net));
    }
    // everything else passes through
    else {
      out.push(op);
    }
  }
  return out;
};



const optimize_clear_loop = (prog) => {
  for (let i = 0; i < prog.length - 2; i++) {
    if (
      prog[i].type === Ops.LOOP_START &&
      ((prog[i + 1].type === Ops.SUB && prog[i + 1].arg === 1) ||
        (prog[i + 1].type === Ops.ADD && prog[i + 1].arg === 1)) &&
      prog[i + 2].type === Ops.LOOP_END
    ) {
      prog.splice(i, 3, new Op(Ops.ZERO));
    }
  }
};


const optimize_copy_loop = (prog) => {
  for (let i = 0; i < prog.length; i++) {
    if (prog[i].type !== Ops.LOOP_START) continue;
    let j = i + 1;
    let pos = 0;
    const copies = [];
    if (!(prog[j]?.type === Ops.SUB && prog[j].arg === 1)) continue;
    j++;
    for (; j < prog.length && prog[j].type !== Ops.LOOP_END; j++) {
      const op = prog[j];
      if (op.type === Ops.RIGHT) pos += op.arg;
      else if (op.type === Ops.LEFT) pos -= op.arg;
      else if (op.type === Ops.ADD) copies.push({ offset: pos, mult: op.arg });
      else break;
    }
    if (pos === 0 && prog[j]?.type === Ops.LOOP_END) {
      const newOps = copies.map(c => new Op(Ops.MULTIPLY, c.mult, c.offset));
      newOps.push(new Op(Ops.ZERO));
      prog.splice(i, j - i + 1, ...newOps);
      i--;
    }
  }
};


const optimize_multiplication_loops = (prog) => {
  outer: for (let i = 0; i < prog.length; i++) {
    if (prog[i].type !== Ops.LOOP_START) continue;

    let j = i + 1;
    let ptrDelta = 0;
    const mult = new Map();

    if (!(prog[j] && prog[j].type === Ops.SUB && prog[j].arg === 1)) continue;
    j++;

    while (j < prog.length && prog[j].type !== Ops.LOOP_END) {
      const op = prog[j];

      switch (op.type) {
        case Ops.LEFT:
          ptrDelta -= op.arg;
          break;

        case Ops.RIGHT:
          ptrDelta += op.arg;
          break;

        case Ops.ADD:
        case Ops.SUB: {
          const off = ptrDelta;
          const delta = (op.type === Ops.ADD ? +op.arg : -op.arg);
          mult.set(off, (mult.get(off) || 0) + delta);
          break;
        }

        default:
          continue outer;
      }
      j++;
    }

    if (prog[j]?.type !== Ops.LOOP_END || ptrDelta !== 0) continue;

    const newOps = [];
    for (const [off, m] of mult.entries()) {
      if (off !== 0 && m !== 0)
        newOps.push(new Op(Ops.MULTIPLY, m, off));
    }
    newOps.push(new Op(Ops.ZERO));

    prog.splice(i, j - i + 1, ...newOps);
    i--;
  }
};



const optimize_offsets = (prog) => {
  const out = [];
  let acc = 0;

  const emitMove = () => {
    if (acc === 0) return;
    out.push(new Op(acc > 0 ? Ops.RIGHT : Ops.LEFT, Math.abs(acc)));
    acc = 0;
  };

  const CAN_OFFSET = new Set([Ops.ADD, Ops.SUB, Ops.ZERO, Ops.OUTPUT, Ops.INPUT]);

  for (const op of prog) {
    switch (op.type) {
      case Ops.LEFT:
        acc -= op.arg;
        break;
      case Ops.RIGHT:
        acc += op.arg;
        break;

      default:
        if (CAN_OFFSET.has(op.type)) {
          out.push(new Op(op.type, op.arg, op.offset + acc));
          break;
        }

        emitMove();
        out.push(op);
    }


    if (op.type === Ops.LOOP_START) emitMove();
  }

  emitMove();
  return out;
};




const align_brackets = (prog) => {
  const stack = [];
  prog.forEach((op, i) => {
    if (op.type === Ops.LOOP_START) stack.push(i);
    else if (op.type === Ops.LOOP_END) {
      const open = stack.pop();
      prog[open].arg = i;
      prog[i].arg = open;
    }
  });
};


const bf_eval = (prog) => {
  const cells = new Uint32Array(1 << 16);
  let cc = 0, pc = 0;
  while (pc < prog.length) {
    const op = prog[pc++];
    switch (op.type) {
      case Ops.LEFT:
        cc -= op.arg;
        break;
      case Ops.RIGHT:
        cc += op.arg;
        break;
      case Ops.ADD:
        cells[cc + op.offset] += op.arg;
        break;
      case Ops.SUB:
        cells[cc + op.offset] -= op.arg;
        break;
      case Ops.ZERO:
        cells[cc + op.offset] = 0;
        break;
      case Ops.LOOP_START:
        if (!cells[cc]) pc = op.arg;
        break;
      case Ops.LOOP_END:
        if (cells[cc]) pc = op.arg;
        break;
      case Ops.MULTIPLY:
        cells[cc + op.offset] += cells[cc] * op.arg;
        break;
      case Ops.OUTPUT:
        process.stdout.write(String.fromCharCode(cells[cc + op.offset]));
        break;
      case Ops.INPUT:
        cells[cc + op.offset] = fs.readSync(
          process.stdin.fd,
          Buffer.alloc(1),
          0,
          1,
          null
        )[0];
        break;
    }
  }
};




// build a reverse lookup from code → name
const OpName = Object.entries(Ops)
  .reduce((m, [name, code]) => { m[code] = name; return m; }, {});

function dumpProg(label, prog) {
  console.log(`\n=== ${label} (${prog.length} ops) ===`);
  console.log(
    prog
      .map(op =>
        op.offset
          ? `${OpName[op.type]}(${op.arg}@${op.offset})`
          : `${OpName[op.type]}(${op.arg})`
      )
      .join('  ')
  );
}






export const run = (bytes) => {
  let prog = create_program(bytes);
  //dumpProg('after create_program', prog);

  prog = optimize_contraction(prog);
  //dumpProg('after optimize_contraction', prog);

  optimize_multiplication_loops(prog);
  //dumpProg('after multiplication loops', prog);


  prog = optimize_contraction(prog);
  //dumpProg('after optimize_contraction', prog);

  optimize_clear_loop(prog);
  //dumpProg('after clear loops', prog);


  
  prog = optimize_contraction(prog);
  //dumpProg('after optimize_contraction', prog);

  prog = optimize_offsets(prog);
  //dumpProg('after optimize_offsets', prog);


  align_brackets(prog);
  //dumpProg('after align_brackets', prog);

  bf_eval(prog);
};

File: interp.hs
{-# LANGUAGE BangPatterns #-}
module Main (main, runBytes) where

import qualified Data.ByteString as BS
import qualified Data.Vector.Unboxed.Mutable as MV
import qualified Data.IntMap.Strict as IM
import Data.Word (Word8)
import System.IO (stdin, stdout, hFlush)
import Data.Bits ((.&.), shiftL)
import Control.Monad.ST (runST)
import Data.STRef


data OpType
  = LEFT_OP
  | RIGHT_OP
  | ADD_OP
  | SUB_OP
  | LOOP_START_OP
  | LOOP_END_OP
  | OUTPUT_OP
  | INPUT_OP
  | ZERO_OP
  | MULTIPLY_OP
  deriving (Eq, Show)

data Op = Op
  { opType   :: !OpType
  , opArg    :: !Int
  , opOffset :: !Int
  } deriving (Eq, Show)

newOp :: OpType -> Int -> Int -> Op
newOp = Op

consecOp :: BS.ByteString -> Int -> Word8 -> (Int, Int)
consecOp bytes i target = go i
  where
    len = BS.length bytes
    go !j
      | j < len && BS.index bytes j == target = go (j + 1)
      | otherwise = (j - 1, j - i)

createProgram :: BS.ByteString -> [Op]
createProgram bytes = go 0 []
  where
    len = BS.length bytes
    go !i acc
      | i >= len = reverse acc
      | otherwise =
          case BS.index bytes i of
            60 -> let (i', count) = consecOp bytes i 60 in go (i' + 1) (newOp LEFT_OP count 0 : acc)
            62 -> let (i', count) = consecOp bytes i 62 in go (i' + 1) (newOp RIGHT_OP count 0 : acc)
            43 -> let (i', count) = consecOp bytes i 43 in go (i' + 1) (newOp ADD_OP count 0 : acc)
            45 -> let (i', count) = consecOp bytes i 45 in go (i' + 1) (newOp SUB_OP count 0 : acc)
            91 -> go (i + 1) (newOp LOOP_START_OP 0 0 : acc)
            93 -> go (i + 1) (newOp LOOP_END_OP 0 0 : acc)
            46 -> go (i + 1) (newOp OUTPUT_OP 0 0 : acc)
            44 -> go (i + 1) (newOp INPUT_OP 0 0 : acc)
            _  -> go (i + 1) acc

optimizeContraction :: [Op] -> [Op]
optimizeContraction = reverse . foldl step []
  where
    signedData op = case opType op of
      ADD_OP -> opArg op
      SUB_OP -> negate (opArg op)
      _      -> 0
    signedPtr op = case opType op of
      RIGHT_OP -> opArg op
      LEFT_OP  -> negate (opArg op)
      _        -> 0

    emitData n acc
      | n > 0     = newOp ADD_OP n 0 : acc
      | n < 0     = newOp SUB_OP (-n) 0 : acc
      | otherwise = acc

    emitPtr n acc
      | n > 0     = newOp RIGHT_OP n 0 : acc
      | n < 0     = newOp LEFT_OP (-n) 0 : acc
      | otherwise = acc

    step [] op
      | opType op == ADD_OP || opType op == SUB_OP = emitData (signedData op) []
      | opType op == LEFT_OP || opType op == RIGHT_OP = emitPtr (signedPtr op) []
      | otherwise = [op]
    step acc@(lastOp:rest) op
      | opType op == ADD_OP || opType op == SUB_OP =
          if opType lastOp == ADD_OP || opType lastOp == SUB_OP
            then emitData (signedData op + signedData lastOp) rest
            else emitData (signedData op) acc
      | opType op == LEFT_OP || opType op == RIGHT_OP =
          if opType lastOp == LEFT_OP || opType lastOp == RIGHT_OP
            then emitPtr (signedPtr op + signedPtr lastOp) rest
            else emitPtr (signedPtr op) acc
      | otherwise = op : acc

optimizeClearLoop :: [Op] -> [Op]
optimizeClearLoop [] = []
optimizeClearLoop [a] = [a]
optimizeClearLoop [a,b] = [a,b]
optimizeClearLoop (a:b:c:rest)
  | opType a == LOOP_START_OP
    && ((opType b == SUB_OP && opArg b == 1) || (opType b == ADD_OP && opArg b == 1))
    && opType c == LOOP_END_OP = newOp ZERO_OP 0 0 : optimizeClearLoop rest
  | otherwise = a : optimizeClearLoop (b:c:rest)

optimizeCopyLoop :: [Op] -> [Op]
optimizeCopyLoop = go
  where
    go [] = []
    go (startOp:rest)
      | opType startOp /= LOOP_START_OP = startOp : go rest
      | otherwise =
          case rest of
            (seed:xs) | opType seed == SUB_OP && opArg seed == 1 ->
              case scanLoop 0 [] xs of
                Just (copies, remaining) ->
                  let newOps = [newOp MULTIPLY_OP mult off | (off, mult) <- copies] ++ [newOp ZERO_OP 0 0]
                  in newOps ++ go remaining
                Nothing -> startOp : go rest
            _ -> startOp : go rest

    scanLoop !_ acc [] = Nothing
    scanLoop !pos acc (op:xs) =
      case opType op of
        RIGHT_OP -> scanLoop (pos + opArg op) acc xs
        LEFT_OP  -> scanLoop (pos - opArg op) acc xs
        ADD_OP   -> scanLoop pos (acc ++ [(pos, opArg op)]) xs
        LOOP_END_OP | pos == 0 -> Just (acc, xs)
        _ -> Nothing

optimizeMultiplicationLoops :: [Op] -> [Op]
optimizeMultiplicationLoops = go
  where
    go [] = []
    go (startOp:rest)
      | opType startOp /= LOOP_START_OP = startOp : go rest
      | otherwise =
          case rest of
            (seed:xs) | opType seed == SUB_OP && opArg seed == 1 ->
              case scanLoop 0 IM.empty xs of
                Just (multMap, remaining) ->
                  let newOps = [newOp MULTIPLY_OP m off | (off, m) <- IM.toList multMap, off /= 0, m /= 0]
                           ++ [newOp ZERO_OP 0 0]
                  in newOps ++ go remaining
                Nothing -> startOp : go rest
            _ -> startOp : go rest

    scanLoop !_ !_ [] = Nothing
    scanLoop !ptrDelta !multMap (op:xs) =
      case opType op of
        LEFT_OP  -> scanLoop (ptrDelta - opArg op) multMap xs
        RIGHT_OP -> scanLoop (ptrDelta + opArg op) multMap xs
        ADD_OP   -> scanLoop ptrDelta (IM.insertWith (+) ptrDelta (opArg op) multMap) xs
        SUB_OP   -> scanLoop ptrDelta (IM.insertWith (+) ptrDelta (negate (opArg op)) multMap) xs
        LOOP_END_OP | ptrDelta == 0 -> Just (multMap, xs)
        _ -> Nothing

optimizeOffsets :: [Op] -> [Op]
optimizeOffsets ops = finalize $ foldl step ([], 0) ops
  where
    canOffset t = t `elem` [ADD_OP, SUB_OP, ZERO_OP, OUTPUT_OP, INPUT_OP]

    emitMove (out, acc)
      | acc == 0 = (out, 0)
      | acc > 0  = (out ++ [newOp RIGHT_OP acc 0], 0)
      | otherwise = (out ++ [newOp LEFT_OP (-acc) 0], 0)

    step (!out, !acc) op =
      case opType op of
        LEFT_OP  -> (out, acc - opArg op)
        RIGHT_OP -> (out, acc + opArg op)
        LOOP_START_OP ->
          let (out', acc') = emitMove (out, acc)
          in (out' ++ [op], acc')
        _ | canOffset (opType op) -> (out ++ [op { opOffset = opOffset op + acc }], acc)
        _ -> let (out', acc') = emitMove (out, acc)
             in (out' ++ [op], acc')

    finalize st = fst (emitMove st)

alignBrackets :: [Op] -> [Op]
alignBrackets ops = map attach [0 .. n - 1]
  where
    n = length ops
    indexed = zip [0..] ops
    pairMap = build indexed [] IM.empty

    build [] [] m = m
    build [] _ _ = error "Unmatched '['"
    build ((i,op):xs) stack m =
      case opType op of
        LOOP_START_OP -> build xs (i:stack) m
        LOOP_END_OP ->
          case stack of
            [] -> error "Unmatched ']'"
            (open:rest) -> build xs rest (IM.insert i open (IM.insert open i m))
        _ -> build xs stack m

    attach i =
      let op = ops !! i
      in case opType op of
           LOOP_START_OP -> op { opArg = IM.findWithDefault 0 i pairMap }
           LOOP_END_OP   -> op { opArg = IM.findWithDefault 0 i pairMap }
           _             -> op

bfEval :: [Op] -> BS.ByteString -> BS.ByteString
bfEval prog inputBytes = runST $ do
  cells <- MV.replicate (1 `shiftL` 16) (0 :: Int)
  inRef <- newSTRef 0
  outRef <- newSTRef ([] :: [Word8])

  let plen = length prog

      readCell idx = MV.read cells idx
      writeCell idx val = MV.write cells idx val
      modCell idx f = do
        cur <- MV.read cells idx
        MV.write cells idx (f cur)

      exec !pc !cc
        | pc >= plen = pure ()
        | otherwise = do
            let op = prog !! pc
                idx = cc + opOffset op
            case opType op of
              LEFT_OP  -> exec (pc + 1) (cc - opArg op)
              RIGHT_OP -> exec (pc + 1) (cc + opArg op)
              ADD_OP   -> modCell idx (+ opArg op) >> exec (pc + 1) cc
              SUB_OP   -> modCell idx (subtract (opArg op)) >> exec (pc + 1) cc
              ZERO_OP  -> writeCell idx 0 >> exec (pc + 1) cc
              LOOP_START_OP -> do
                cur <- readCell cc
                if cur == 0 then exec (opArg op + 1) cc else exec (pc + 1) cc
              LOOP_END_OP -> do
                cur <- readCell cc
                if cur /= 0 then exec (opArg op) cc else exec (pc + 1) cc
              MULTIPLY_OP -> do
                base <- readCell cc
                modCell idx (+ (base * opArg op))
                exec (pc + 1) cc
              OUTPUT_OP -> do
                cur <- readCell idx
                modifySTRef' outRef (fromIntegral (cur .&. 255) :)
                exec (pc + 1) cc
              INPUT_OP -> do
                ip <- readSTRef inRef
                let b = if ip < BS.length inputBytes then BS.index inputBytes ip else 0
                writeSTRef inRef (ip + 1)
                writeCell idx (fromIntegral b)
                exec (pc + 1) cc

  exec 0 0
  out <- readSTRef outRef
  pure (BS.pack (reverse out))

runBytes :: BS.ByteString -> BS.ByteString
runBytes bytes =
  let prog0 = createProgram bytes
      prog1 = optimizeContraction prog0
      prog2 = optimizeMultiplicationLoops prog1
      prog3 = optimizeContraction prog2
      prog4 = optimizeClearLoop prog3
      prog5 = optimizeContraction prog4
      prog6 = optimizeOffsets prog5
      prog7 = alignBrackets prog6
  in bfEval prog7 BS.empty

main :: IO ()
main = do
  src <- BS.hGetContents stdin
  BS.hPut stdout (runBytes src)
  hFlush stdout
